pyccel/parser/syntactic.py

```diff
@@ -633,54 +643,127 @@ def _visit_FunctionDef(self, stmt):
+                                    symbol = comb_types,
+                                    bounding_box = (stmt.lineno, stmt.col_offset),
+                                    severity='fatal')
+
+                i = 0
+                while i<len(ls):
+                    arg = ls[i]
+                    if isinstance(arg, Symbol):
+                        arg = arg.name
+                        types.append(arg)
+                    elif isinstance(arg, String):
+                        arg = str(arg)
+                        arg = arg.strip("'").strip('"')
+                        types.append(arg)
+                    else:
+                        msg = 'Invalid argument type of type {} passed to template decorator'.format(type(arg))
+                        errors.report(msg,
+                                    symbol = comb_types,
+                                    bounding_box = (stmt.lineno, stmt.col_offset),
+                                    severity='error')
+
+                    i += 1
+
+                txt  = '#$ header template ' + str(tp_name)
+                txt += '(' + '|'.join(types) + ')'
+                if tp_name in templates:
+                    msg = 'The template "{}" is duplicated'.format(tp_name)
+                    errors.report(msg,
+                                bounding_box = (stmt.lineno, stmt.col_offset),
+                                severity='warning')
+
+                templates[tp_name] = hdr_parse(stmts=txt)
+
+        # extract the types to construct a header
+        if 'types' in decorators:
+            if not isinstance(decorators['types'], list):
+                decorators['types'] = [decorators['types']]
+            for comb_types in decorators['types']:
+                types = []
+                results = []
+                container = types
+                i = 0
+                ls = comb_types.args
+
+                while i<len(ls) :
+                    arg = ls[i]
+
+                    if isinstance(arg, Symbol):
+                        arg = arg.name
+                        container.append(arg)
+                    elif isinstance(arg, String):
+                        arg = str(arg)
+                        arg = arg.strip("'").strip('"')
+                        container.append(arg)
+                    elif isinstance(arg, ValuedArgument):
+                        arg_name = arg.name
+                        arg  = arg.value
+                        container = results
+                        if not arg_name == 'results':
+                            msg = 'Argument "{}" provided to the types decorator is not valid'.format(arg_name)
+                            errors.report(msg,
+                                        symbol = comb_types,
+                                        bounding_box = (stmt.lineno, stmt.col_offset),
+                                        severity='error')
+                        else:
```

Comment:
What can the `arg_name` be other than `results`?

Reply:
This is if the user gave the wrong keyword to the types decorators, we raise an error for this

Reply:
Ah sorry, I didn't read this properly, it's the opposite to what I thought